pkg/bpf/probes/probes.go

```diff
@@ -0,0 +1,241 @@
+		return nil, fmt.Errorf("could not run bpftool: %s", err)
+	}
+	if err := json.Unmarshal(out, &features); err != nil {
+		return nil, fmt.Errorf("could not parse bpftool output: %s", err)
+	}
+	return &ProbeManager{features: features}, nil
+}
+
+// SystemConfigProbes performs a check of kernel configuration parameters. It
+// returns an error when parameters required by Cilium are not enabled. It logs
+// warnings when optional parameters are not enabled.
+func (p *ProbeManager) SystemConfigProbes() error {
+	config := p.features.SystemConfig
+	if !config.ConfigBpf.Enabled() {
+		return fmt.Errorf("CONFIG_BPF kernel parameter is required")
+	}
+	if !config.ConfigBpfSyscall.Enabled() {
+		return fmt.Errorf(
+			"CONFIG_BPF_SYSCALL kernel parameter is required")
+	}
+	if !config.ConfigNetSchIngress.Enabled() && !config.ConfigNetSchIngress.Module() {
+		return fmt.Errorf(
+			"CONFIG_NET_SCH_INGRESS kernel parameter (or module) is required")
+	}
+	if !config.ConfigNetClsBpf.Enabled() && !config.ConfigNetClsBpf.Module() {
+		return fmt.Errorf(
+			"CONFIG_NET_CLS_BPF kernel parameter (or module) is required")
+	}
+	if !config.ConfigNetClsAct.Enabled() {
+		return fmt.Errorf(
+			"CONFIG_NET_CLS_ACT kernel parameter is required")
+	}
+	if !config.ConfigBpfJit.Enabled() {
+		return fmt.Errorf(
+			"CONFIG_BPF_JIR kernel parameter is required")
+	}
+	if !config.ConfigHaveEbpfJit.Enabled() {
+		return fmt.Errorf(
+			"CONFIG_HAVE_EBPF_JIT kernel parameter is required")
+	}
+	if !config.ConfigCgroupBpf.Enabled() {
+		log.Warning(
+			"CONFIG_CGROUP_BPF kernel parameter is not in kernel configuration")
+	}
+	if !config.ConfigLwtunnelBpf.Enabled() {
+		log.Warning(
+			"CONFIG_LWTUNNEL_BPF kernel parameter is not in kernel configuration")
+	}
+	if !config.ConfigBpfEvents.Enabled() {
+		log.Warning(
+			"CONFIG_BPF_EVENTS kernel parameter is not in kernel configuration")
+	}
+	return nil
+}
+
+// GetSupportedMapTypes returns information about supported BPF map types.
+func (p *ProbeManager) GetSupportedMapTypes() *SupportedMapTypes {
+	return &p.features.MapTypes
+}
+
+func generateHeaders(featuresFile io.Writer) error {
+	cmd := exec.WithTimeout(
+		defaults.ExecTimeout, "bpftool", "feature", "probe", "macros")
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf(
+			"could not initialize stout pipe for bpftool: %s", err)
+	}
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("could not start bpftool: %s", err)
+	}
+
+	writer := bufio.NewWriter(featuresFile)
+	defer writer.Flush()
+
+	go io.Copy(writer, stdoutPipe)
+	if err := cmd.Wait(); err != nil {
+		return fmt.Errorf("bpftool did not run successfully: %s", err)
+	}
+	return nil
+}
+
+// CreateHeadersFile creates a C header file with macros indicating which BPF
+// features are available in the kernel.
+func CreateHeadersFile() error {
+	globalsDir := option.Config.GetGlobalsDir()
+	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil {
```

Comment:
Can `err` be `os.IsExist()` in this case? Then we won't want to `Fatal()` out.

Reply:
You are right.